File: 2021/2021W23/PreppindataW23.py
import pandas as pd

xlsx = pd.ExcelFile("F:/Data/NPS Input.xlsx")

temp = []
for sheet in xlsx.sheet_names:
    df = pd.read_excel(xlsx,sheet)
    temp.append(df)
like = pd.concat(temp)   
like.columns = ['Airline', 'CustomerID', 'like']
like['RespCount'] = like['CustomerID'].groupby(like['Airline']).transform('count')
like = like[like['RespCount']>=50].copy()

like['Class'] = like['like'].apply(lambda x: 'Detractors' if x<=6 else 'Passive' if x<=8 else 'Promoters') 

final = like.pivot_table(index='Airline', columns='Class', values='CustomerID', aggfunc='count').reset_index()
final['Promoters%'] = final.apply(lambda x: int(x['Promoters']/(x['Promoters']+x['Passive']+x['Detractors'])*100), axis=1)
final['Detractors%'] = final.apply(lambda x: int(x['Detractors']/(x['Promoters']+x['Passive']+x['Detractors'])*100), axis=1)
final['NPS'] = final['Promoters%'] - final['Detractors%']
final['Z-Score'] = round((final['NPS']-final['NPS'].mean())/final['NPS'].std(),2)
final = final[final['Airline']=='Prep Air'][['Airline', 'NPS', 'Z-Score']].copy()

File: 2021/2021W23/preppindataW23.R
library(readxl)
library(dplyr)
library(tidyr)
library(purrr)

input <- "F:/Data/NPS Input.xlsx"
final <-  input %>%
  excel_sheets() %>% 
  set_names() %>%
  map_df(~ read_excel(path = input, sheet = .x,)) %>%
  `colnames<-`(c('Airline', 'CustomerID', 'like')) %>%
  group_by(`Airline`) %>%
  filter(n()>=50) %>%
  ungroup() %>%
  mutate('Class' = if_else(`like`<=6,'Detractors',if_else(`like`<=8,'Passive','Promoters'))) %>%
  pivot_wider(id_cols='Airline', names_from='Class', values_from='CustomerID', values_fn=list(`CustomerID` = length)) %>%
  mutate('Promoters%' = as.integer(`Promoters`/(`Promoters`+`Passive`+`Detractors`)*100),
         'Detractors%' = as.integer(`Detractors`/(`Promoters`+`Passive`+`Detractors`)*100),
         'NPS' = `Promoters%` - `Detractors%`,
         'Z-Score' = round((`NPS`-mean(`NPS`))/sd(`NPS`),2)) %>%
  filter(`Airline`=='Prep Air') %>%
  select(c('Airline', 'NPS', 'Z-Score'))

View(final)
